feat(slo): Handle updating SLO

kdelemme opened the pull request:
## 📝 Summary

Related to https://github.com/elastic/kibana/issues/141283

This PR introduces the slo revision and uses the tuple (slo.id, slo.revision) when creating the related Transform and aggregating the data.
It also introduces the route to update an existing slo along the application service handling the update.

As explained in the related [story](https://github.com/elastic/kibana/issues/141283), an edit can be breaking or non breaking. 
For non-breaking changes, i.e. editing anything but the indicator: we only need to update the saved object without bumping the slo revision.
For breaking changes, i.e. changing the indicator: we need to remove the previously aggregated data and transform (could become optional later with an extra parameter if we want to retain the previous data), create a new transform and start it.


## ✅ Manual testing


1. Create a new slo
2. Check revision in kibana saved object, expect revision === 1
3. Update the slo without breaking changes (change the name, time_window, objective...)
4. Get the slo 
5. Assert
   1. SLO data has been updated
   2. Transform is still running
   3. Check revision in kibana saved object, expect revision === 1
3. Update the slo with breaking changes (change the indicator's environment for example)
4. Get the slo 
5. Assert
   1. SLO data has been updated, including indicator
   2. Previous transform (named `slo-{id}-1`) has been deleted
   3. New transform (named `slo-{id}-2`) is running
   4. Check revision in kibana saved object, expect revision === 2
 


### ⚙️  Useful requests

<details><summary>Create SLO</summary>

```
curl --request POST \
  --url http://localhost:5601/mgx/api/observability/slos \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'kbn-xsrf: oui' \
  --data '{
	"name": "My SLO Availability",
	"description": "99% o11y-app all services availablility",
	"indicator": {
		"type": "slo.apm.transaction_error_rate",
		"params": {
			"environment": "development",
			"service": "o11y-app",
			"transaction_type": "request",
			"transaction_name": "GET /flaky",
			"good_status_codes": ["2xx", "3xx"]
		}
	},
	"time_window": {
		"duration": "7d",
		"is_rolling": true
	},
	"budgeting_method": "occurrences",
	"objective": {
		"target": 0.95
	}
}'
```
</details>


<details><summary>Update SLO (without breaking changes)</summary>

```
curl --request PUT \
  --url http://localhost:5601/mgx/api/observability/slos/7a42fe70-4022-11ed-93fe-659ac997b883 \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'kbn-xsrf: oui' \
  --data '{
	"name": "My Updated SLO",
	"time_window": {
		"duration": "30d",
		"is_rolling": true
	}
}'
```
</details>


<details><summary>Update SLO (with breaking changes)</summary>

```
curl --request PUT \
  --url http://localhost:5601/mgx/api/observability/slos/7a42fe70-4022-11ed-93fe-659ac997b883 \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'kbn-xsrf: oui' \
  --data '{
	"name": "Updated SLO bis",
	"indicator": {
		"type": "slo.apm.transaction_error_rate",
		"params": {
			"environment": "production",
			"service": "o11y-app",
			"transaction_type": "*",
			"transaction_name": "GET /flaky",
			"good_status_codes": [
				"2xx",
				"3xx",
				"4xx"
			]
		}
	}
}'
```
</details>



<details><summary>Get SLO</summary>

```
curl --request GET \
  --url http://localhost:5601/mgx/api/observability/slos/7a42fe70-4022-11ed-93fe-659ac997b883 \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'kbn-xsrf: oui'
```
</details>

kdelemme committed:
Improve types

Introduce date type schema and a revision value for the slo model.

kdelemme committed:
Introduce slo revision

kdelemme committed:
Use indicator type in generators

kdelemme committed:
Handle update of SLO

elasticmachine commented:
Pinging @elastic/actionable-observability (Team: Actionable Observability)

maryam-saeidi commented on x-pack/plugins/observability/server/services/slo/get_slo.ts:
```diff
@@ -26,6 +26,9 @@ export class GetSLO {
       budgeting_method: slo.budgeting_method,
       objective: slo.objective,
+      revision: slo.revision,
+      created_at: slo.created_at,
```
What has changed that now you return `created_at` and `updated_at` fields?

kdelemme replied:
I've thought a bit more about the UI, and used the Alerts page as example, and I think we'll have similar display and sorting functionalities, which includes sorting per creation/update time for example

maryam-saeidi commented on x-pack/plugins/observability/server/services/slo/slo_repository.ts:
```diff
@@ -22,24 +22,18 @@ export class KibanaSavedObjectsSLORepository implements SLORepository {
-    );
+    const savedSLO = await this.soClient.create<StoredSLO>(SO_SLO_TYPE, slo, {
+      id: slo.id,
+      overwrite: true,
```
Is `overwrite: true` for the update scenario?

maryam-saeidi reviewed (commented):
Tested locally and everything worked as expected! ⭐ 

I've added some clarification questions. In general, the update logic was clear but I also see some refactoring in this PR which I am not sure why you did the changes, can you please explain your thought process for the refactoring part? (Regarding types and namings)

Usually, it is easier to review PRs when refactoring and new logic are separated into different PRs.

kdelemme commented on x-pack/plugins/observability/server/routes/slo/route.ts:
```diff
@@ -53,6 +54,26 @@ const createSLORoute = createObservabilityServerRoute({
+    const soClient = (await context.core).savedObjects.client;
+
+    const repository = new KibanaSavedObjectsSLORepository(soClient);
+    const transformManager = new DefaultTransformManager(transformGenerators, esClient, logger);
```
You'll notice I like OOP (but I'll try to include more functional code from time to time when it makes sense 😅 ), therefore I usually define an interface and one or many classes. 
When I don't have a better name for the main implementation, I name it `Default<InterfaceName>`. Usually, the name includes the specificities about the implementation, e.g. `Repository` (interface) with the following implementations: `SqlRepository` as opposed to `InMemoryRepository` for example.

maryam-saeidi replied:
I see, so in this case, `DefaultTransformManager` is like a conversion not that necessary there is a plan to have another implementation for `TransformManager`, right?

kdelemme replied:
Technically I have some dummy implementations in the tests 😅 

https://github.com/kdelemme/kibana/blob/feat/141283-update-slo/x-pack/plugins/observability/server/services/slo/transform_manager.test.ts#L156-L166

kdelemme commented:
> I've added some clarification questions. In general, the update logic was clear but I also see some refactoring in this PR which I am not sure why you did the changes, can you please explain your thought process for the refactoring part? (Regarding types and namings)

Not much changed in fact, but there is a lot of moving part I must confess. I basically reintroduce duplication in the rest_specs, so each route/application service types can be defined without relying on a `commonSLO` base schema which might or might not be so common between the route. I think it's easier this way to understand what is the specs for each route in one sight.
The StoredSLO is currently identical to the slo model, that's why I simplified the model there, but this might change as we support new feature and want to store extra things (like history of changes, alerts execution, etc...) and that's still possible to reintroduce.
